cpp/src/arrow/compute/exec/spilling_util.h

```diff
@@ -0,0 +1,96 @@
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <atomic>
+#include <vector>
+#include <functional>
+#include <memory>
+#include "arrow/compute/exec/query_context.h"
+
+namespace arrow
+{
+    namespace compute
+    {
+#ifdef _WIN32
+        using FileHandle = void *;
+        extern const FileHandle kInvalidHandle;
+#else
+        using FileHandle = int;
+        constexpr FileHandle kInvalidHandle = -1;
+#endif
+
+        // A temporary file meant for spilling data to disk. It can spill a batch to
+        // disk and read it back into memory. This class is designed to fully utilize
+        // disk bandwidth and for removing batches from memory as quickly as possible.
+        class SpillFile
+        {
+        public:
+            static constexpr size_t kAlignment = 512;
+
+            ~SpillFile();
+            // To spill a batch the following must be true:
+            // - Row offset for each column must be 0.
+            // - Column buffers must be aligned to 512 bits
+            // - No column can be a scalar
+            // These assumptions aren't as inconvenient as it seems because
+            // typically batches will be partitioned before being spilled,
+            // meaning the batches will come from ExecBatchBuilder, which
+            // ensures these assumptions hold. 
+            // It is a bug to spill a batch after ReadBackBatches.
+            Status SpillBatch(QueryContext *ctx, ExecBatch batch);
+
+            // Reads back all of the batches from the disk, invoking `fn`
+            // on each batch, and invoking `on_finished` when reading back
+            // is done. For example, `fn` could be used to insert into an
+            // AccumulationQueue for further processing. Note that `fn` will
+            // be run on IO thread pool, so it should only be used to park
+            // the ExecBatch somewhere until `on_finished` is called on the
+            // compute thread pool. 
+            Status ReadBackBatches(
+                QueryContext *ctx,
+                std::function<Status(size_t, ExecBatch)> fn,
+                std::function<Status(size_t)> on_finished);
+            Status Cleanup();
+            size_t num_batches() const { return batches_.size(); }
+            size_t batches_written() const { return batches_written_.load(); }
+
+            // Used for benchmarking only!
+            Status PreallocateBatches(MemoryPool *memory_pool);
+
+            struct BatchInfo;
+        private:
+            Status ScheduleReadbackTasks(QueryContext *ctx);
+
+            bool preallocated_ = false;
+
+            FileHandle handle_ = kInvalidHandle;
+            size_t size_ = 0;
+
+            std::vector<BatchInfo *> batches_;
+
+            std::atomic<size_t> batches_written_{0};
+            std::atomic<bool> read_requested_{false};
+            std::atomic<bool> read_started_{false};
+            std::atomic<size_t> batches_read_{0};
```

Comment:
There is a complicated access pattern for that class on which the correctness of checks depends. Public methods are called by a worker thread after taking an exclusive lock, so only one worker thread at a time will be using an instance of this class. On the other hand IO threads may be making updates concurrently. Correctness of thread synchronization right now relies on understanding the above, but this is far from obvious when reading the code in this file and also some expressions are not correct if we consider multi-threaded access by both worker and IO threads. Would be good to comment about that to make it easy to reason about correctness of statements involving access to these variables.

Reply:
Added a comment above the class.